// browser/navigation-panel.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

define((require, exports, module) => {

'use strict';

const {DOM} = require('react')
const Component = require('omniscient');
const {InputField} = require('./editable');
const {Element} = require('./element');
const {navigateTo, showTabStrip, focus} = require('./actions');
const {KeyBindings} = require('./keyboard');
const url = require('./util/url');

const WindowControls = Component(({webViewer}) => {
  const theme = webViewer.get('theme');
  const closeColor = theme.get('close');
  const minColor = theme.get('min');
  const maxColor = theme.get('max');
  const glyphsShowing = theme.get('glyphsShowing');

  return DOM.div({className: 'windowctrls' +
                  (glyphsShowing ? ' glyphsshowing' : '')}, [
    DOM.div({className: 'windowctrl win-close-button',
             style: { backgroundColor: closeColor },
             title: 'close',
             key: 'close',
             onClick: event => window.close()}),
    DOM.div({className: 'windowctrl win-min-button',
             style: { backgroundColor: minColor },
             title: 'minimze',
             key: 'minimize',
             onClick: event => window.minimize()}),
    DOM.div({className: 'windowctrl win-max-button',
             style: { backgroundColor: maxColor },
             title: 'maximize',
             key: 'maximize',
             onClick: event => {
               if (document.mozFullScreenElement) {
                 document.mozCancelFullScreen();
               } else {
                 document.body.mozRequestFullScreen();
               }
             }})
  ])});
exports.WindowControls = WindowControls;

const inputBindings = KeyBindings({'escape': focus});


const NavigationControls = Component(({input, webViewer, title}) => {
  const theme = webViewer.get('theme');
  const buttonColor = theme.get('buttons');
  const titleColor = theme.get('title');
  const domainColor = theme.get('domain');

  const location = webViewer.get('location')

  return DOM.div({
    className: 'locationbar',
    onMouseEnter: event => showTabStrip(input)
    }, [
    DOM.div({className: 'backbutton',
             style: {color: buttonColor},
             key: 'back',
             onClick: event => webViewer.set('readyState', 'goBack')}),
    InputField({
      key: 'input',
      className: 'urlinput',
      style: {color: domainColor},
      placeholder: 'Search or enter address',
      value: input.get('value') || webViewer.get('location'),
      type: 'text',
      submitKey: 'Enter',
      isFocused: input.get('isFocused'),
      selection: input.get('isFocused'),
      onFocus: _ => input.set('isFocused', true),
      onBlur: _ => input.set('isFocused', false),
      onChange: event => input.set('value', event.target.value),
      onSubmit: event => navigateTo({input, webViewer}, event.target.value, true),
      onKeyUp: inputBindings(webViewer),
    }),
    DOM.p({key: 'page-info',
           className: 'pagesummary',
           style: {color: titleColor},
           onClick: event => input.set('isFocused', true)}, [
      DOM.span({key: 'identity',
                style: {color: buttonColor},
                className: 'identity'}, ''),
      DOM.span({key: 'location',
                style: {color: domainColor},
                className: 'pageurlsummary'},
                location ? url.getDomainName(location) : ''),
      DOM.span({key: 'title',
                style: {color: titleColor},
                className: 'pagetitle'},
               title ? title :
               webViewer.get('isLoading') ? 'Loading...' :
               location ? location :
               'New Tab')
    ]),
    DOM.div({key: 'reload-button',
             className: 'reloadbutton',
             style: {color: buttonColor},
             onClick: event => webViewer.set('readyState', 'reload')}),
    DOM.div({key: 'stop-button',
             className: 'stopbutton',
             style: {color: buttonColor},
             onClick: event => webViewer.set('readyState', 'stop')})
  ]);
});
exports.NavigationControls = NavigationControls;

const NavigationPanel = Component(({key, input, webViewer, title}) => {
  const theme = webViewer.get('theme');
  const isDark = theme.get('isDark');
  const backgroundColor = theme.get('background');

  return DOM.div({
    key,
    style: { backgroundColor: backgroundColor },
    className: 'navbar' + (input.get('isFocused') ? ' urledit' : '')
                        + (isDark ? " isdark" : '')
                        + (webViewer.get('canGoBack') ? ' cangoback' : '')
                        + (webViewer.get('location') ? ' canreload' : '')
                        + (webViewer.get('isLoading') ? ' loading' : '')
                        + (webViewer.get('securityState') == 'secure' ? ' ssl' : '')
                        + (webViewer.get('securityExtendedValidation') ? ' sslv' : '')
  }, [
    WindowControls({key: 'controls', webViewer}),
    NavigationControls({key: 'navigation', input, webViewer, title}),
    DOM.div({key: 'spacer', className: 'freeendspacer'})
  ])
  });
exports.NavigationPanel = NavigationPanel;

});
